tests/ts_cluster_aae.erl

```diff
@@ -0,0 +1,404 @@
+                                    get_aae_throttle_limits,
+                                    []))
+      end,
+      Nodes).
+
+verify_aae_ts(Nodes) ->
+    DDL = ts_data:get_ddl(),
+    ts_setup:create_bucket_type(Nodes, DDL, ?TABLE),
+    ts_setup:activate_bucket_type(Nodes, ?TABLE),
+
+    Node1 = hd(Nodes),
+
+    % Verify that AAE eventually upgrades to version 0(or already has)
+    wait_until_hashtree_upgrade(Nodes), 
+    
+    % Recovery without tree rebuilds
+
+    % Test recovery from to few replicas written
+    KV1 = test_data_ts(1, 1000),
+    test_less_than_n_writes_ts(Node1, KV1, ?TABLE, ?N_VAL, <<"weather">>, [{n_val, 1}]),
+
+    % Test recovery when replicas are different
+    KV2 = [{A, B, C, <<D/binary, "a">>, E} || {A, B, C, D, E} <- KV1],
+    test_less_than_n_mods_ts(Node1, KV2, ?TABLE, ?N_VAL, <<"weather">>, [{n_val, 1}]),
+
+    lager:info("Run similar tests now with tree rebuilds enabled"),
+    start_tree_rebuilds_ts(Nodes),
+
+    % Test recovery from too few replicas written
+    KV3 = test_data_ts(1001, 2000),
+    test_less_than_n_writes_ts(Node1, KV3, ?TABLE, ?N_VAL, <<"weather">>, [{n_val, 1}]),
+
+    % Test recovery when replicas are different
+    KV4 = [{A, B, C, <<D/binary, "a">>, E} || {A, B, C, D, E} <- KV3],
+    test_less_than_n_mods_ts(Node1, KV4, ?TABLE, ?N_VAL, <<"weather">>, [{n_val, 1}]),
+
+    lager:info("Writing 1000 objects"),
+    KV5 = test_data_ts(2001, 3000),
+    write_data_ts(Node1, KV5, ?TABLE),
+
+    % Test recovery from single partition loss.
+    {PNuke, NNuke} = choose_partition_to_nuke_ts(Node1, ?TABLE, KV5, ?N_VAL),
+    test_single_partition_loss_ts(NNuke, PNuke, KV5, ?TABLE, ?N_VAL, <<"weather">>),
+
+    % Test recovery from losing AAE data
+    test_aae_partition_loss_ts(NNuke, PNuke, KV5, ?TABLE, ?N_VAL, <<"weather">>),
+
+    % Test recovery from losing both AAE and KV data
+    test_total_partition_loss_ts(NNuke, PNuke, KV5, ?TABLE, ?N_VAL, <<"weather">>),
+
+    % Make sure AAE repairs die down.
+    wait_until_no_aae_repairs(Nodes),
+
+    lager:info("Finished verifying TS AAE magic"),
+    ok.
+
+start_tree_rebuilds_ts(Nodes) ->
+    rpc:multicall(Nodes, application, set_env, [riak_kv, anti_entropy_expire,
+                                                15 * 1000]).
+
+acc_preflists(Pl, PlCounts) ->
+    lists:foldl(fun(Idx, D) ->
+                        dict:update(Idx, fun(V) -> V+1 end, 0, D)
+                end, PlCounts, Pl).
+
+test_data_ts(Start, End) ->
+    StepSize = 1000*60*5,
+    ts_data:get_valid_select_data(fun() -> lists:seq(1 + StepSize*Start, 1 + StepSize*End, StepSize) end).
+
+key_ts(Data) -> 
+    lists:sublist(tuple_to_list(Data), 3).
+
+value_ts(Data) -> 
+    lists:sublist(tuple_to_list(Data), 4, 10).
+
+write_data_ts(Node, KVs, Table) ->
+    write_data_ts(Node, KVs, Table, []).
+
+write_data_ts(Node, KVs, Table, _Opts) ->
+    [begin
+         O =
+         case ts_ops:get([Node], Table, key_ts(TS_Data_Tuple)) of
```

Comment:
Why are we reading the data first? In the `verify_aae` case they read in case the data already exists and update it instead.  In TS, it is write-once and we never update values.